File: Reaktoro/Core/ChemicalSystem-test.py
# Reaktoro is a unified framework for modeling chemically reactive systems.
#
# Copyright (C) 2014-2020 Allan Leal
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this library. If not, see <http://www.gnu.org/licenses/>.

from reaktoro import *
from pytest import approx, raises
from numpy import array


def names(objects):
    """Extract the names of each object in a list."""
    return [o.name() for o in objects]


def range_species_in_phase(system, iphase):
    """Return the range of species indices in a phase."""
    return range(system.indexFirstSpeciesInPhase(iphase),
        system.indexFirstSpeciesInPhase(iphase) + system.numSpeciesInPhase(iphase))


def test_chemical_system():
    """Test function for class ChemicalSystem."""

    editor = ChemicalEditor()
    editor.addAqueousPhase("H2O(l) H+ OH- HCO3- CO2(aq) CO3--".split())
    editor.addGaseousPhase("H2O(g) CO2(g)".split())
    editor.addMineralPhase("Graphite")
    system = ChemicalSystem(editor)

    # The number of elements, species, and phases
    Ne = system.numElements()
    Ns = system.numSpecies()
    Np = system.numPhases()

    # A sensible value for temperature (in K)
    T = 300

    # A sensible value for pressure (in Pa)
    P = 1e5

    # A sensible array of species amounts
    n = array([55, 1e-7, 1e-7, 0.1, 0.5, 0.01, 1.0, 0.001, 1.0])

    # The formula matrix of the system
    A = system.formulaMatrix()

    # The amounts of the elements in the system
    b = A @ n

    # The formula matrices of each phase in the system
    Ap = [ A[:, range_species_in_phase(system, j)] for j in range(Np) ]

    # The amounts of the species of each phase in the system
    np = [ n[range_species_in_phase(system, j)] for j in range(Np) ]

    # The amounts of the elements of each phase in the system
    bp = [ Ap[j] @ np[j] for j in range(Np) ]

    # Calculate the chemical properties of the state
    properties = system.properties(T, P, n)

    # -------------------------------------------------------------------------
    # Check methods ChemicalSystem::num(Elements|Species|Phases)
    # -------------------------------------------------------------------------
    assert Ne == len(system.elements())
    assert Ns == len(system.species())
    assert Np == len(system.phases())

    # -------------------------------------------------------------------------
    # Check method ChemicalSystem::numSpeciesInPhase
    # -------------------------------------------------------------------------
    for j in range(Np):
        assert system.numSpeciesInPhase(j) == len(system.phase(j).species())

    # -------------------------------------------------------------------------
    # Check method ChemicalSystem::formulaMatrix
    # -------------------------------------------------------------------------
    assert system.formulaMatrix() == approx(array([
        [ s.elementCoefficient(e.name()) for s in system.species()]
            for e in system.elements() ]))

    # -------------------------------------------------------------------------
    # Check method ChemicalSystem::element
    # -------------------------------------------------------------------------
    for i, element in enumerate(system.elements()):
        assert system.element(i) == element

    for i, element in enumerate(system.elements()):
        assert system.element(element.name()) == element

    # -------------------------------------------------------------------------
    # Check method ChemicalSystem::species
    # -------------------------------------------------------------------------
    for i, species in enumerate(system.species()):
        assert system.species(i) == species

    # Check method ChemicalSystem::species(name)
    for species in system.species():
        assert system.species(species.name()) == species

    # -------------------------------------------------------------------------
    # Check method ChemicalSystem::phase
    # -------------------------------------------------------------------------
    for i, phase in enumerate(system.phases()):
        assert system.phase(i) == phase

    for i, phase in enumerate(system.phases()):
        assert system.phase(phase.name()) == phase

    # -------------------------------------------------------------------------
    # Check method ChemicalSystem::indexElement
    # -------------------------------------------------------------------------
    for i, element in enumerate(system.elements()):
        assert system.indexElement(element.name()) == i

    for i, element in enumerate(system.elements()):
        assert system.indexElementWithError(element.name()) == i

    with raises(RuntimeError):
        assert system.indexElementWithError("Aa")

    # -------------------------------------------------------------------------
    # Check method ChemicalSystem::indexSpecies
    # -------------------------------------------------------------------------
    for i, species in enumerate(system.species()):
        assert system.indexSpecies(species.name()) == i

    for i, species in enumerate(system.species()):
        assert system.indexSpeciesWithError(species.name()) == i

    with raises(RuntimeError):
        assert system.indexSpeciesWithError("AaBb2")

    # -------------------------------------------------------------------------
    # Check method ChemicalSystem::indexSpeciesAny
    # -------------------------------------------------------------------------
    for i in range(Ns):
        name = system.species(i).name()
        assert system.indexSpeciesAny(["AaBb", name]) == system.indexSpecies(name)

    with raises(RuntimeError):
        assert system.indexSpeciesAnyWithError(["AaBb2", "AaCc3"])

    # -------------------------------------------------------------------------
    # Check method ChemicalSystem::indexPhase
    # -------------------------------------------------------------------------
    for i, phase in enumerate(system.phases()):
        assert system.indexPhase(phase.name()) == i

    for i, phase in enumerate(system.phases()):
        assert system.indexPhaseWithError(phase.name()) == i

    with raises(RuntimeError):
        assert system.indexPhaseWithError("WrongName")

    # -------------------------------------------------------------------------
    # Check method ChemicalSystem::indexPhaseWithSpecies
    # -------------------------------------------------------------------------
    for iphase, phase in enumerate(system.phases()):
        for ispecies, species in enumerate(phase.species()):
            i = system.indexSpecies(species.name())
            assert system.indexPhaseWithSpecies(i) == iphase

    # -------------------------------------------------------------------------
    # Check method ChemicalSystem::indexFirstSpeciesInPhase
    # -------------------------------------------------------------------------
    for iphase, phase in enumerate(system.phases()):
        i = system.indexSpecies(phase.species(0).name())
        assert system.indexFirstSpeciesInPhase(iphase) == i

    # -------------------------------------------------------------------------
    # Check method ChemicalSystem::indicesElements
    # -------------------------------------------------------------------------
    assert system.indicesElements(names(system.elements())) == approx(range(Ne))

    # -------------------------------------------------------------------------
    # Check method ChemicalSystem::indicesElementsInSpecies
    # -------------------------------------------------------------------------
    for ispecies, species in enumerate(system.species()):
        ielements = [system.indexElement(e.name()) for e in species.elements()]
        assert system.indicesElementsInSpecies(ispecies) == approx(ielements)

    # -------------------------------------------------------------------------
    # Check method ChemicalSystem::indicesSpecies
    # -------------------------------------------------------------------------
    assert system.indicesSpecies(names(system.species())) == approx(range(Ns))

    # -------------------------------------------------------------------------
    # Check method ChemicalSystem::indicesSpeciesInPhases
    # -------------------------------------------------------------------------
    assert system.indicesSpeciesInPhases(range(Np)) == approx(range(Ns))

    for iphase, phase in enumerate(system.phases()):
        assert system.indicesSpeciesInPhases([iphase]) == \
            system.indicesSpecies(names(phase.species()))

    # -------------------------------------------------------------------------
    # Check method ChemicalSystem::indicesPhases
    # -------------------------------------------------------------------------
    assert system.indicesPhases(names(system.phases())) == approx(range(Np))

    # -------------------------------------------------------------------------
    # Check method ChemicalSystem::indicesPhasesWithSpecies
    # -------------------------------------------------------------------------
    assert system.indicesPhasesWithSpecies(array(range(Ns))) == approx(range(Np))

    # -------------------------------------------------------------------------
    # Check method ChemicalSystem::indicesFluidPhases
    # -------------------------------------------------------------------------
    assert system.indicesFluidPhases() == \
        approx([i for i, phase in enumerate(system.phases()) if phase.isFluid()])

    # -------------------------------------------------------------------------
    # Check method ChemicalSystem::indicesFluidSpecies
    # -------------------------------------------------------------------------
    assert system.indicesFluidSpecies() == \
        approx(system.indicesSpeciesInPhases(system.indicesFluidPhases()))

    # -------------------------------------------------------------------------
    # Check method ChemicalSystem::indicesSolidPhases
    # -------------------------------------------------------------------------
    assert system.indicesSolidPhases() == \
        approx([i for i, phase in enumerate(system.phases()) if phase.isSolid()])

    # -------------------------------------------------------------------------
    # Check method ChemicalSystem::indicesSolidSpecies
    # -------------------------------------------------------------------------
    assert system.indicesSolidSpecies() == \
        approx(system.indicesSpeciesInPhases(system.indicesSolidPhases()))

    # -------------------------------------------------------------------------
    # Check method ChemicalSystem::elementAmounts
    # -------------------------------------------------------------------------
    assert system.elementAmounts(n) == approx(b)

    # -------------------------------------------------------------------------
    # Check method ChemicalSystem::elementAmountsInPhase
    # -------------------------------------------------------------------------
    for j in range(Np):
        assert system.elementAmountsInPhase(j, n) == approx(bp[j])

    # -------------------------------------------------------------------------
    # Check method ChemicalSystem::elementAmountsInSpecies
    # -------------------------------------------------------------------------
    assert system.elementAmountsInSpecies([], n) == approx(0.0)
    assert system.elementAmountsInSpecies(range(Ns), n) == approx(b)

    # -------------------------------------------------------------------------
    # Check method ChemicalSystem::elementAmount
    # -------------------------------------------------------------------------
    for j in range(Ne):
        assert system.elementAmount(j, n) == approx(b[j])

    # -------------------------------------------------------------------------
    # Check method ChemicalSystem::elementAmountInPhase
    # -------------------------------------------------------------------------
    for ielement, iphase in ((x, y) for x in range(Ne) for y in range(Np)):
        assert system.elementAmountInPhase(ielement, iphase, n) == approx(bp[iphase][ielement])

    # -------------------------------------------------------------------------
    # Check method ChemicalSystem::elementAmountInSpecies
    # -------------------------------------------------------------------------
    for ielement in range(Ne):
        assert system.elementAmountInSpecies(ielement, [], n) == approx(0.0)
        assert system.elementAmountInSpecies(ielement, range(Ns), n) == approx(b[ielement])

    # Check the usage system.properties(T, P, n).someProperty() works
    assert all(system.properties(T, P, n).phaseVolumes().val == properties.phaseVolumes().val)
    assert all(system.properties(T, P, n).lnActivities().val == properties.lnActivities().val)
    assert all(system.properties(T, P, n).chemicalPotentials().val == properties.chemicalPotentials().val)

File: Reaktoro/Core/ChemicalSystem.test.cxx
// Reaktoro is a unified framework for modeling chemically reactive systems.
//
// Copyright © 2014-2022 Allan Leal
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 2.1 of the License, or (at your option) any later version.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with this library. If not, see <http://www.gnu.org/licenses/>.

// Catch includes
#include <catch2/catch.hpp>

// Reaktoro includes
#include <Reaktoro/Core/ChemicalSystem.hpp>
#include <Reaktoro/Core/Utils.hpp>
using namespace Reaktoro;

namespace test {

/// Return a mock Database object for test reasons (imported).
extern auto createDatabase() -> Database;

/// Return mock activity properties for an aqueous solution.
auto activityModelAqueous(ActivityPropsRef props, ActivityModelArgs args)
{
    const auto [T, P, x] = args;
    props.Vx  = 0.1 * log(P/T);
    props.VxT = 0.2 * log(P/T);
    props.VxP = 0.3 * log(P/T);
    props.Gx  = 0.4 * log(P/T);
    props.Hx  = 0.5 * log(P/T);
    props.Cpx = 0.6 * log(P/T);
    props.ln_g = 0.8 * x;
    props.ln_a = 0.9 * x;
};

/// Return mock activity properties for a gaseous solution.
auto activityModelGaseous(ActivityPropsRef props, ActivityModelArgs args)
{
    const auto [T, P, x] = args;
    props.Vx  = 1.0 * log(P/T);
    props.VxT = 2.0 * log(P/T);
    props.VxP = 3.0 * log(P/T);
    props.Gx  = 4.0 * log(P/T);
    props.Hx  = 5.0 * log(P/T);
    props.Cpx = 6.0 * log(P/T);
    props.ln_g = 8.0 * x;
    props.ln_a = 9.0 * x;
};

/// Return mock activity properties for a solid solution.
auto activityModelSolid(ActivityPropsRef props, ActivityModelArgs args)
{
    const auto [T, P, x] = args;
    props.Vx  = 1.1 * log(P/T);
    props.VxT = 2.1 * log(P/T);
    props.VxP = 3.1 * log(P/T);
    props.Gx  = 4.1 * log(P/T);
    props.Hx  = 5.1 * log(P/T);
    props.Cpx = 6.1 * log(P/T);
    props.ln_g = 8.1 * x;
    props.ln_a = 9.1 * x;
};

/// Return a Phase object representing an aqueous phase using mock thermodynamic models.
auto createAqueousPhase(const Database& db) -> Phase
{
    Phase phase;
    phase = phase.withName("AqueousPhase");
    phase = phase.withSpecies(db.speciesWithAggregateState(AggregateState::Aqueous));
    phase = phase.withStateOfMatter(StateOfMatter::Liquid);
    phase = phase.withActivityModel(activityModelAqueous);
    phase = phase.withIdealActivityModel(ActivityModelIdealAqueous()(phase.species()));
    return phase;
}

/// Return a Phase object representing a gaseous phase using mock thermodynamic models.
auto createGaseousPhase(const Database& db) -> Phase
{
    Phase phase;
    phase = phase.withName("GaseousPhase");
    phase = phase.withSpecies(db.speciesWithAggregateState(AggregateState::Gas));
    phase = phase.withStateOfMatter(StateOfMatter::Gas);
    phase = phase.withActivityModel(activityModelGaseous);
    phase = phase.withIdealActivityModel(ActivityModelIdealGas()(phase.species()));
    return phase;
}

/// Return a Phase object representing a mineral phase using mock thermodynamic models.
auto createMineralPhase(const Database& db, String phasename, String mineral) -> Phase
{
    Phase phase;
    phase = phase.withName(phasename);
    phase = phase.withSpecies({ db.species().get(mineral) });
    phase = phase.withStateOfMatter(StateOfMatter::Solid);
    phase = phase.withActivityModel(activityModelSolid);
    phase = phase.withIdealActivityModel(ActivityModelIdealSolution(StateOfMatter::Solid)(phase.species()));
    return phase;
}

/// Return a mock ChemicalSystem object for test reasons.
auto createChemicalSystem() -> ChemicalSystem
{
    // Create the Database object for the ChemicalSystem
    Database db = test::createDatabase();

    // Create the Phase objects for the ChemicalSystem
    const PhaseList phases = {
        createAqueousPhase(db),
        createGaseousPhase(db),
        createMineralPhase(db, "Halite", "NaCl(s)"),
        createMineralPhase(db, "Calcite", "CaCO3(s)"),
        createMineralPhase(db, "Magnesite", "MgCO3(s)"),
        createMineralPhase(db, "Quartz", "SiO2(s)"),
    };

    const ReactionList reactions = {
        db.reaction("H2O(aq) = H+(aq) + OH-(aq)"),
        db.reaction("H2O(aq) = H2(aq) + 0.5*O2(aq)"),
        db.reaction("O2(g) = O2(aq)"),
        db.reaction("NaCl(s) = Na+(aq) + Cl-(aq)"),
        db.reaction("CaCO3(s)"),
    };

    return ChemicalSystem(db, phases, reactions);
}

} // namespace test

TEST_CASE("Testing ChemicalSystem class", "[ChemicalSystem]")
{
    //-------------------------------------------------------------------------
    // TESTING CONSTRUCTOR: ChemicalSystem::ChemicalSystem(phases)
    //-------------------------------------------------------------------------
    ChemicalSystem system = test::createChemicalSystem();

    //-------------------------------------------------------------------------
    // TESTING METHOD: ChemicalSystem::id()
    //-------------------------------------------------------------------------
    CHECK( ChemicalSystem().id() == system.id() + 1 ); // new ChemicalSystem object has id = id0 + 1 and original system continues to have id = id0
    CHECK( ChemicalSystem().id() == system.id() + 2 ); // new ChemicalSystem object has id = id0 + 2 and original system continues to have id = id0

    //-------------------------------------------------------------------------
    // TESTING METHOD: ChemicalSystem::database()
    //-------------------------------------------------------------------------
    CHECK( system.database().species().size() );

    //-------------------------------------------------------------------------
    // TESTING METHOD: ChemicalSystem::phases()
    //-------------------------------------------------------------------------
    CHECK( system.phases().size() == 6 );

    //-------------------------------------------------------------------------
    // TESTING METHOD: ChemicalSystem::phase(i)
    //-------------------------------------------------------------------------
    CHECK( system.phase(0).name() == "AqueousPhase" );
    CHECK( system.phase(1).name() == "GaseousPhase" );
    CHECK( system.phase(2).name() == "Halite"       );
    CHECK( system.phase(3).name() == "Calcite"      );
    CHECK( system.phase(4).name() == "Magnesite"    );
    CHECK( system.phase(5).name() == "Quartz"       );

    //-------------------------------------------------------------------------
    // TESTING METHOD: ChemicalSystem::species()
    //-------------------------------------------------------------------------
    CHECK( system.species().size() == 29 );

    //-------------------------------------------------------------------------
    // TESTING METHOD: ChemicalSystem::species(i)
    //-------------------------------------------------------------------------
    CHECK( system.species(0).name()  == "H2O(aq)"   );
    CHECK( system.species(1).name()  == "H+(aq)"    );
    CHECK( system.species(2).name()  == "OH-(aq)"   );
    CHECK( system.species(3).name()  == "H2(aq)"    );
    CHECK( system.species(4).name()  == "O2(aq)"    );
    CHECK( system.species(5).name()  == "Na+(aq)"   );
    CHECK( system.species(6).name()  == "Cl-(aq)"   );
    CHECK( system.species(7).name()  == "NaCl(aq)"  );
    CHECK( system.species(8).name()  == "HCl(aq)"   );
    CHECK( system.species(9).name()  == "NaOH(aq)"  );
    CHECK( system.species(10).name() == "Ca++(aq)"  );
    CHECK( system.species(11).name() == "Mg++(aq)"  );
    CHECK( system.species(12).name() == "CO2(aq)"   );
    CHECK( system.species(13).name() == "HCO3-(aq)" );
    CHECK( system.species(14).name() == "CO3--(aq)" );
    CHECK( system.species(15).name() == "CaCl2(aq)" );
    CHECK( system.species(16).name() == "MgCl2(aq)" );
    CHECK( system.species(17).name() == "SiO2(aq)"  );
    CHECK( system.species(18).name() == "e-(aq)"    );
    CHECK( system.species(19).name() == "CO2(g)"    );
    CHECK( system.species(20).name() == "O2(g)"     );
    CHECK( system.species(21).name() == "H2(g)"     );
    CHECK( system.species(22).name() == "H2O(g)"    );
    CHECK( system.species(23).name() == "CH4(g)"    );
    CHECK( system.species(24).name() == "CO(g)"     );
    CHECK( system.species(25).name() == "NaCl(s)"   );
    CHECK( system.species(26).name() == "CaCO3(s)"  );
    CHECK( system.species(27).name() == "MgCO3(s)"  );
    CHECK( system.species(28).name() == "SiO2(s)"   );

    //-------------------------------------------------------------------------
    // TESTING METHOD: ChemicalSystem::elements()
    //-------------------------------------------------------------------------
    CHECK( system.elements().size() == 8 );

    //-------------------------------------------------------------------------
    // TESTING METHOD: ChemicalSystem::element(i)
    //-------------------------------------------------------------------------
    Strings symbols = { "H", "O", "Na", "Cl", "C", "Ca", "Mg", "Si" };

    CHECK( contains(symbols, system.element(0).symbol()) );
    CHECK( contains(symbols, system.element(1).symbol()) );
    CHECK( contains(symbols, system.element(2).symbol()) );
    CHECK( contains(symbols, system.element(3).symbol()) );
    CHECK( contains(symbols, system.element(4).symbol()) );
    CHECK( contains(symbols, system.element(5).symbol()) );
    CHECK( contains(symbols, system.element(6).symbol()) );
    CHECK( contains(symbols, system.element(7).symbol()) );

    //-------------------------------------------------------------------------
    // TESTING METHOD: ChemicalSystem::reactions()
    //-------------------------------------------------------------------------
    CHECK( system.reactions().size() == 5 );

    //-------------------------------------------------------------------------
    // TESTING METHOD: ChemicalSystem::reaction(i)
    //-------------------------------------------------------------------------
    CHECK( system.reaction(0).name() == "H2O(aq) = H+(aq) + OH-(aq)" );
    CHECK( system.reaction(1).name() == "H2O(aq) = H2(aq) + 0.5*O2(aq)" );
    CHECK( system.reaction(2).name() == "O2(g) = O2(aq)" );
    CHECK( system.reaction(3).name() == "NaCl(s) = Na+(aq) + Cl-(aq)" );
    CHECK( system.reaction(4).name() == "CaCO3(s)" );

    //-------------------------------------------------------------------------
    // TESTING METHOD: ChemicalSystem::formulaMatrix()
    //-------------------------------------------------------------------------

    const auto Aexpected = detail::assembleFormulaMatrix(system.species(), system.elements());

    CHECK(system.formulaMatrix() == Aexpected);
    CHECK(system.formulaMatrixElements() == Aexpected.topRows(system.elements().size()));
    CHECK(system.formulaMatrixCharge() == Aexpected.row(system.elements().size()));

    //-------------------------------------------------------------------------
    // TESTING METHOD: ChemicalSystem::stoichiometricMatrix()
    //-------------------------------------------------------------------------

    const auto Sexpected = detail::assembleStoichiometricMatrix(system.reactions(), system.species());

    CHECK(system.stoichiometricMatrix() == Sexpected);

    //-------------------------------------------------------------------------
    // TESTING METHOD: ChemicalSystem::surfaces()
    //-------------------------------------------------------------------------

    auto surfaces = system.surfaces();

    CHECK( surfaces.size() == 3 ); // aqueous-gaseous, aqueous-halite, calcite-implicit phase

    CHECK( surfaces[0].name()         == "AqueousPhase:GaseousPhase" );
    CHECK( surfaces[0].phaseNames()   == Pair<String, String>{ "AqueousPhase", "GaseousPhase" } );
    CHECK( surfaces[0].phaseIndices() == Pair<Index, Index>{ 0, 1 } );

    CHECK( surfaces[1].name()         == "AqueousPhase:Halite" );
    CHECK( surfaces[1].phaseNames()   == Pair<String, String>{ "AqueousPhase", "Halite" } );
    CHECK( surfaces[1].phaseIndices() == Pair<Index, Index>{ 0, 2 } );

    CHECK( surfaces[2].name()         == "Calcite" );
    CHECK( surfaces[2].phaseNames()   == Pair<String, String>{ "Calcite", "Calcite" } );
    CHECK( surfaces[2].phaseIndices() == Pair<Index, Index>{ 3, 3 } );

    //-------------------------------------------------------------------------
    // TESTING METHOD: ChemicalSystem::surfaces().indexWithPhases()
    //-------------------------------------------------------------------------

    CHECK( system.surfaces().findWithPhases("AqueousPhase", "GaseousPhase") == 0 );
    CHECK( system.surfaces().findWithPhases("AqueousPhase", "Halite") == 1 );
    CHECK( system.surfaces().findWithPhases("Calcite", "Calcite") == 2 );

    CHECK( system.surfaces().findWithPhases("Calcite", "Halite") >= system.surfaces().size() );
    CHECK( system.surfaces().findWithPhases("Calcite", "Quartz") >= system.surfaces().size() );

    //-------------------------------------------------------------------------
    // TESTING CONSTRUCTOR: ChemicalSystem::ChemicalSystem(db, phases...)
    //-------------------------------------------------------------------------
    auto db = test::createDatabase();
    system = ChemicalSystem(db,
        AqueousPhase("H2O(aq) H+(aq) OH-(aq) Na+(aq) Cl-(aq)"),
        GaseousPhase("H2O(g)"),
        MineralPhase("NaCl(s)"),
        MineralPhases("CaCO3(s) MgCO3(s) SiO2(s)"));

    CHECK( system.elements().size() == 8 );
    CHECK( system.species().size() == 10 );
    CHECK( system.phases().size() == 6 );
}
